pulpcore/pulpcore/app/models/repository.py

```diff
@@ -221,25 +207,147 @@ class RepositoryContent(Model):
 
         content (models.ForeignKey): The associated content.
         repository (models.ForeignKey): The associated repository.
+        vadded (models.ForeignKey): The RepositoryVersion which added the referenced Content.
+        vremoved (models.ForeignKey): The RepositoryVersion which removed the referenced Content.
     """
-    content = models.ForeignKey('Content', on_delete=models.CASCADE)
+    content = models.ForeignKey('Content', on_delete=models.CASCADE,
+                                related_name='version_memberships')
     repository = models.ForeignKey(Repository, on_delete=models.CASCADE)
+    vadded = models.ForeignKey('RepositoryVersion', related_name='added_memberships')
+    vremoved = models.ForeignKey('RepositoryVersion', null=True, related_name='removed_memberships')
+
+    class Meta:
+        unique_together = (('repository', 'content', 'vadded'),
+                           ('repository', 'content', 'vremoved'))
+
+
+class RepositoryVersion(Model):
+    """
+    A version of a repository's content set.
+    Fields:
+        number (models.PositiveIntegerField): A positive integer that uniquely identifies a version
+            of a specific repository. Each new version for a repo should have this field set to
+            1 + the most recent version.
+        created (models.DateTimeField): When the version was created.
+        action  (models.TextField): The action that produced the version.
+    Relations:
+        repository (models.ForeignKey): The associated repository.
+    """
+    SYNC = 'sync'
+    COPY = 'copy'
+    UPLOAD = 'upload'
+    ADD = 'add'
+    REMOVE = 'remove'
+    ACTIONS = (
+        (SYNC, 'Sync'),
+        (COPY, 'Copy'),
+        (UPLOAD, 'Upload'),
+        (ADD, 'Add'),
+        (REMOVE, 'Remove'),
+    )
+
+    repository = models.ForeignKey(Repository)
```

Comment:
I think this needs a user_visible field that defaults to False. It is set to true as the last step of the a sync before recording it on the task status as the 'created_resource'. This is written in the MVP use cases some. This resolves the inconsistant state comments below. Note that the viewset should *not* display user_visible; to the user it doesn't exist.

Reply:
Also the MVP talks about having a 'add_many' and a 'remove_many' both of which would be list href attributes to be added and removed in bulk via a POST to the RepositoryContent detail. To add them I think 3 things need to happen:

1. They need to be recorded at the DB level as data provided by the user for the creation of this repo version.
2. During the sync, after the creation of the repo version, core should associate the add_many units and remove the remove_many units before calling into any specific importer.

Reply:
I like that idea. We've talked at some point about a similar thing for publications also.

There's also a use case for recovery from an incomplete operation, for example a sync that crashed. I think it would go like this: When the core is about to make a new version, it would look for any existing versions with this boolean set to ``False`` and delete them. Then it would proceed with making a new version and handing it to the appropriate plugin code.

Any other thoughts on that?

Reply:
Regarding the "add_many" and "remove_many" comment, I'm just not following. Can you clarify? Did you mean to say "RepositoryVersion" instead of "RepositoryContent"? What's the user story behind this? Is it "As a user, I can add and remove content in a repository by POSTing the representation of a new RepositoryVersion to the repo version collection endpoint" or something like that?

Reply:
+1 to the design you propose regarding deletion. That is a better design because it makes Pulp more crash-safe. I think we are planning to do the exact same thing in Publications (from the MVP call convo).

Reply:
Yes I mean "RepositoryVersion" not "RepositoryComment" I updated the comment.

Sure. The use case idea is to have this fulfil both move and copy and associate/unassociate. For instance you upload and create N content units, you could get them into a repo with `add_many=['href1', 'href2', 'href3']` etc. Or copy these Y units into a repo, or remove these Z units, etc. It's immutable so the only thing the user can do is to add and/or remove content when making a new one anyway. Another option they can use in conjunction w/ add_many, remove_many is to call an importer and have it import content. It's a strange use case, but the idea is the options could all be used together.

Reply:
> During the sync, after the creation of the repo version, core should associate the add_many units and remove the remove_many units before calling into any specific importer.

Are you suggesting a user would supply "add_many" and "remove_many" to a sync operation? I don't see how those would work together. I would expect those two options to be parameters for a different kind of task besides a sync.